runtime/major_gc.c

```diff
@@ -368,119 +517,82 @@ static value* mark_ephe_aux (value *gray_vals_ptr, intnat *work,
     *ephes_checked_if_pure = v;
     ephes_checked_if_pure = &Field(v,CAML_EPHE_LINK_OFFSET);
   }
-  return gray_vals_ptr;
 }
 
 
 
 static void mark_slice (intnat work)
 {
-  value *gray_vals_ptr;  /* Local copy of [gray_vals_cur] */
-  value v;
-  header_t hd;
-  mlsize_t size, i, start, end; /* [start] is a local copy of [current_index] */
+  mark_entry me = {0, 0};
+  mlsize_t me_end = 0;
 #ifdef CAML_INSTR
   int slice_fields = 0; /** eventlog counters */
 #endif /*CAML_INSTR*/
   int slice_pointers = 0;
+  struct mark_stack* stk = Caml_state->mark_stack;
 
   caml_gc_message (0x40, "Marking %"ARCH_INTNAT_PRINTF_FORMAT"d words\n", work);
   caml_gc_message (0x40, "Subphase = %d\n", caml_gc_subphase);
-  gray_vals_ptr = gray_vals_cur;
-  v = current_value;
-  start = current_index;
-  while (work > 0){
-    if (v == 0 && gray_vals_ptr > gray_vals){
-      CAMLassert (start == 0);
-      v = *--gray_vals_ptr;
-      CAMLassert (Is_gray_val (v));
-#ifdef NO_NAKED_POINTERS
-      if (Tag_val(v) == Closure_tag) {
-        /* Skip the code pointers and integers at beginning of closure;
-           start scanning at the first word of the environment part. */
-        start = Start_env_closinfo(Closinfo_val(v));
-        CAMLassert(start <= Wosize_val(v));
+
+  while (1){
+    int can_mark = 0;
```

Comment:
I'm grateful for the simplification in this "suspendable / restartable" loop, compared with the original code.  The micro-optimization person in me still wonders whether the `can_mark` Boolean could be replaced by control flow (a `goto` or two).  Maybe not.

Reply:
I'm also grateful for the simplification and the use of `can_mark` looks OK to me.